Automattic/newspack-theme
# Feat: Add sticky header

### All Submissions:


### Changes proposed in this Pull Request:

This PR adds an option to make the header sticky. It uses a CSS-only approach with `position: sticky`, and a touch of AMP-Animation/JS to help it work with the behind/beside featured image placements and the simplified subpage header.

After some experimentation, `position: sticky` seemed the most solid, and required the least 'song and dance' to make sure it worked nicely with different header sizes and settings, but I'm open to feedback on the approach!

**Potential issues:**

1. As part of this change, I moved the `overflow: hidden` from the `#page` element to the `.site-content` element. This overlow style makes sure the full-width blocks don't cause a slight side scroll, but with the way `position: sticky` works, the direct parent of the element with this style can't have an overflow style set. This change is applied whether or not the header is sticky, and is the only change that is; I don't anticipate any issues, but wanted to flag it in case someone else did. 
2. The theme enqueues the `amp-animation` and `amp-position-observer` script; I'm not sure if this'll clash (somehow) with another plugin like Campaigns, and if there's anything special that needs to be done. I didn't see any issues in my testing, but that doesn't mean anything!
3. This option adds a fixed smaller size to the site logo, so the header can't get too large.
4. The 'Highlight Menu' is hidden via CSS for all of these options; it looked super strange out of the box, but I didn't want to fully remove it in case a publisher wanted to style it to make it work. 
5. `position: sticky` doesn't have full support, but should fail gracefully; the header will not be sticky in IE11 or Opera Mini: https://caniuse.com/?search=position%3A%20sticky.

Closes #84.

### How to test the changes in this Pull Request:

Apologies in advance for how involved these testing steps are! 

1. Start with the default theme, and a test environment where you can turn on ads (alternatively, you can use the above and below widget areas - just need to test whether a header that's pushed down 'behaves' predictably). 
2. Apply the PR and run `npm run build`.
3. Assign a menu to all locations (Primary, Secondary, Tertiary, Social, Topic Highlights).
4. Navigate to Customize > Header Settings > Appearance, and uncheck the first three options.
5. Navigate to Customize > Header Settings > Slide-out Sidebar, and enable it.
6. View on the front-end - your header should look roughly like this:
![image](https://user-images.githubusercontent.com/177561/109245127-c346e300-7794-11eb-9bc4-051aed5a8ac4.png)

7. Navigate back to Customize > Header Settings > Appearance, and check the new 'Sticky' option.
8. View on the front-end and scroll; confirm that the header sticks, and the borders for the header extend the width of the browser window, and the 'Highlights' menu is hidden:

![image](https://user-images.githubusercontent.com/177561/109243806-6ea26880-7792-11eb-9fb1-1d302d1a0421.png)

9. Confirm that the Slide-out sidebar opens, even after scrolling.
10. Scale down the browser window to check the tablet and mobile screen sizes, and confirm that the sticky header still works, and that the mobile menu opens fine, even after scrolling:

![image](https://user-images.githubusercontent.com/177561/109243853-84179280-7792-11eb-92a7-66c40ca4fbea.png)

11. Set up three posts: with the image behind, image beside and image above image placements; view and make sure the header is sticky, and no weird display issues appear (like gaps between the header and image), on desktop and smaller screen sizes:

![image](https://user-images.githubusercontent.com/177561/109245393-3e0ffe00-7795-11eb-8c65-1ab37388b5a3.png)

![image](https://user-images.githubusercontent.com/177561/109245829-0786b300-7796-11eb-9a6a-7b271e5edf8f.png)

![image](https://user-images.githubusercontent.com/177561/109245399-42d4b200-7795-11eb-9336-8a03ad624e70.png)

12. Navigate to Customize > Header Settings > Subpage Header, and check "Use simple header on subpages". 
13. View the featured image behind and beside again; confirm the header is sticky, and when you start scrolling, a dark background appears behind it:

![image](https://user-images.githubusercontent.com/177561/109246038-61877880-7796-11eb-82d7-6b1a89d0f6fc.png)

![image](https://user-images.githubusercontent.com/177561/109246049-66e4c300-7796-11eb-9321-25571f18c15c.png)

14. Toggle AMP (off if you currently have it on, and on if you currently have it off); repeat the previous step to confirm the dark background still appears.
15. Set up a centre or top overlay campaign, and test with a couple pages; confirm that there are no z-index issues with the sticky header:

![image](https://user-images.githubusercontent.com/177561/109253118-eaf17780-77a3-11eb-95e3-e4d0b4215033.png)

![image](https://user-images.githubusercontent.com/177561/109253436-aa462e00-77a4-11eb-9ead-dea9121c051c.png)

16. Set up an above header inline campaign, and confirm that the sticky header works as expected (remaining static until the campaign is scrolled out of sight):

![image](https://user-images.githubusercontent.com/177561/109253562-eaa5ac00-77a4-11eb-8938-d65903a64671.png)

![stickyheader 2021-02-25 20_10_45](https://user-images.githubusercontent.com/177561/109253868-93eca200-77a5-11eb-9494-c01a2597ceb9.gif)

17. Test with an above-header ad:

![image](https://user-images.githubusercontent.com/177561/109254402-cfd43700-77a6-11eb-837c-f1ae8793629a.png)

![image](https://user-images.githubusercontent.com/177561/109254183-5a686680-77a6-11eb-8e4a-648f88d9f49a.png)

18. Test with a below-header ad (which should not be visible with the beside and behind image options):

![image](https://user-images.githubusercontent.com/177561/109255216-69e8af00-77a8-11eb-9ba6-ac7eef6eaa6b.png)

![image](https://user-images.githubusercontent.com/177561/109256424-d4025380-77aa-11eb-8ed1-bf6b0f658833.png)

19. Switch the theme to Newspack Nelson.
20. Retest the regular sticky header, and the simplified subpage header with regular posts, and a post with each the behind and beside featured image placement. The 'overlap' header style should be overwritten, and the bottom border of the header should pick up the primary colour:

![image](https://user-images.githubusercontent.com/177561/109257672-80ddd000-77ad-11eb-8311-e96cea7bbc07.png)

![image](https://user-images.githubusercontent.com/177561/109257984-33159780-77ae-11eb-9007-8a9d15639332.png)

21. Switch the theme to Newspack Joseph.
22. Retest the regular sticky header, and the simplified subpage header with regular posts, and a post with each the behind and beside featured image placement. The 'frame' around the body isn't maintained on scroll (would be interested to know if this feels 'broken':

![image](https://user-images.githubusercontent.com/177561/109258137-812a9b00-77ae-11eb-8d30-9a1530fab70b.png)

### Other information:
